audio/out/ao_avfoundation.m

```diff
@@ -0,0 +1,354 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ao.h"
+#include "audio/format.h"
+#include "audio/out/ao_coreaudio_chmap.h"
+#include "audio/out/ao_coreaudio_utils.h"
+#include "common/common.h"
+#include "common/msg.h"
+#include "internal.h"
+#include "osdep/timer.h"
+#include "ta/ta_talloc.h"
+
+#import <AVFoundation/AVFoundation.h>
+#import <Foundation/Foundation.h>
+#import <CoreAudioTypes/CoreAudioTypes.h>
+#import <CoreFoundation/CoreFoundation.h>
+#import <CoreMedia/CoreMedia.h>
+
+
+@interface AVObserver : NSObject {
+    struct ao *ao;
+}
+- (void)handleRestartNotification:(NSNotification*)notification;
+@end
+
+struct priv {
+    AVSampleBufferAudioRenderer *renderer;
+    AVSampleBufferRenderSynchronizer *synchronizer;
+    dispatch_queue_t queue;
+    CMAudioFormatDescriptionRef format_description;
+    AVObserver *observer;
+    int64_t end_time_av;
+};
+
+static int64_t CMTimeGetNanoseconds(CMTime time)
+{
+    time = CMTimeConvertScale(time, 1000000000, kCMTimeRoundingMethod_Default);
+    return time.value;
+}
+
+static CMTime CMTimeFromNanoseconds(int64_t time)
+{
+    return CMTimeMake(time, 1000000000);
+}
+
+static void feed(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+    int samplerate = ao->samplerate;
+    int sstride = ao->sstride;
+
+    CMBlockBufferRef block_buffer = NULL;
+    CMSampleBufferRef sample_buffer = NULL;
+    OSStatus err;
+
+    int request_sample_count = samplerate / 10;
+    int buffer_size = request_sample_count * sstride;
+    void *data[] = {CFAllocatorAllocate(NULL, buffer_size, 0)};
+
+    int64_t cur_time_av = CMTimeGetNanoseconds([p->synchronizer currentTime]);
+    int64_t cur_time_mp = mp_time_ns();
+    int64_t end_time_av = MPMAX(p->end_time_av, cur_time_av);
+    int64_t time_delta = CMTimeGetNanoseconds(CMTimeMake(request_sample_count, samplerate));
+    int real_sample_count = ao_read_data_nonblocking(ao, data, request_sample_count, end_time_av - cur_time_av + cur_time_mp + time_delta);
```

Comment:
How about another API change to make this simpler?

```diff
// The caller should set out_time_us to the expected delay until the last sample
// reaches the speakers, in microseconds, using mp_time_us() as reference.
+// Alternatively, out_time_us can be a negative number indicating the delay in
+// microseconds, the absolute value of which will be added to mp_time_us().
```

Reply:
IMO it is counterintuitive to have different meanings for positive values and negative values.